File: ios/Music/ObsidianMusicPlayerModule.swift
import AVFoundation
import Foundation
import MediaPlayer

@objc(ObsidianMusicPlayer)
class ObsidianMusicPlayer: RCTEventEmitter {
  private let player = AVPlayer()
  private let remote = ObsidianRemoteControls()
  private var tracks: [MusicTrack] = []
  private var order: [Int] = []
  private var cursor = 0
  private var repeatMode = "off"   // off | track | queue
  private var shuffle = false
  private var listeners = Set<String>()
  private var timeObserver: Any?

  private(set) var lastState: [String: Any] = [
    "status": "idle", "position": 0, "duration": 0, "rate": 1,
    "muted": false, "volume": 1, "buffered": 0, "inBackground": false,
  ]

  override init() {
    super.init()
    remote.onCommand = { [weak self] cmd in self?.handleRemoteCommand(cmd) }
    remote.onSeek = { [weak self] pos in self?.player.seek(to: CMTime(seconds: pos, preferredTimescale: 1000)) }
    timeObserver = player.addPeriodicTimeObserver(
      forInterval: CMTime(seconds: 0.25, preferredTimescale: 1000),
      queue: .main
    ) { [weak self] time in
      guard let self = self else { return }
      self.lastState["position"] = max(0, time.seconds)
      self.lastState["duration"] = self.player.currentItem?.duration.seconds ?? 0
      self.delegateProgress()
    }
    NotificationCenter.default.addObserver(
      self, selector: #selector(itemEnded),
      name: .AVPlayerItemDidPlayToEndTime, object: nil
    )
  }

  @objc override static func requiresMainQueueSetup() -> Bool { false }

  override func supportedEvents() -> [String] {
    ["onState", "onProgress", "onTrackChange", "onQueueChange", "onRemoteCommand", "onEnded", "onError"]
  }

  @objc func addListener(_ eventName: String) { listeners.insert(eventName) }
  @objc func removeListeners(_ count: Double) {
    if Int(count) >= listeners.count { listeners.removeAll() }
  }

  private func emit(_ name: String, _ body: [String: Any]) {
    if listeners.contains(name) || listeners.isEmpty { sendEvent(withName: name, body: body) }
  }

  // MARK: - Queue

  private func rebuildOrder(preserveCurrent: Bool = true) {
    let n = tracks.count
    order = Array(0..<n)
    if shuffle, n > 1 {
      for i in (1..<n).reversed() {
        let j = Int.random(in: 0...i)
        order.swapAt(i, j)
      }
      if preserveCurrent, n > 0 {
        let active = currentIndex()
        if let pos = order.firstIndex(of: active), pos != 0 {
          order.swapAt(0, pos)
        }
      }
    }
    if cursor >= order.count { cursor = 0 }
  }

  private func currentIndex() -> Int { order.isEmpty ? -1 : order[cursor] }

  private func loadCurrent(autoPlay: Bool = false) {
    guard !order.isEmpty else { return }
    let idx = order[cursor]
    let track = tracks[idx]
    let asset = AVURLAsset(
      url: URL(string: track.source.uri) ?? URL(fileURLWithPath: track.source.uri),
      options: track.source.headers.map { ["AVURLAssetHTTPHeaderFieldsKey": $0] } ?? [:]
    )
    let item = AVPlayerItem(asset: asset)
    item.addObserver(self, forKeyPath: #keyPath(AVPlayerItem.status), context: nil)
    item.addObserver(self, forKeyPath: #keyPath(AVPlayerItem.loadedTimeRanges), context: nil)
    player.replaceCurrentItem(with: item)
    lastState["status"] = "loading"
    updateNowPlaying()
    emit("onTrackChange", ["index": idx, "trackJson": trackJson(at: idx)])
    emit("onState", ["stateJson": stateJson()])
    if autoPlay { player.play() }
  }

  @objc func setQueue(_ tracksJson: String) {
    tracks = decodeTracks(tracksJson)
    rebuildOrder(preserveCurrent: false)
    cursor = 0
    loadCurrent()
    emitQueue()
  }
  @objc func addTracks(_ tracksJson: String) {
    tracks.append(contentsOf: decodeTracks(tracksJson))
    rebuildOrder()
    emitQueue()
  }
  @objc func removeTrack(_ id: String) {
    tracks.removeAll { $0.id == id }
    rebuildOrder()
    emitQueue()
    loadCurrent()
  }
  @objc func skipTo(_ index: Double) {
    let target = Int(index)
    if let pos = order.firstIndex(of: target) { cursor = pos }
    else if target < tracks.count { cursor = max(0, order.firstIndex(of: target) ?? 0) }
    loadCurrent()
  }
  @objc func next() {
    if order.isEmpty { return }
    if cursor < order.count - 1 { cursor += 1 }
    else if repeatMode == "queue" { cursor = 0 }
    else { lastState["status"] = "ended"; emit("onEnded", [:]); return }
    loadCurrent(autoPlay: true)
  }
  @objc func previous() {
    if order.isEmpty { return }
    if cursor > 0 { cursor -= 1 }
    else if repeatMode == "queue" { cursor = order.count - 1 }
    loadCurrent(autoPlay: true)
  }

  // MARK: - Transport

  @objc func play() { configureSession(); player.play() }
  @objc func pause() { player.pause() }
  @objc func stop() { player.pause(); player.seek(to: .zero) }
  @objc(seek:) func seek(_ seconds: Double) { player.seek(to: CMTime(seconds: seconds, preferredTimescale: 1000)) }
  @objc(setRate:) func setRate(_ rate: Double) { player.rate = Float(rate); lastState["rate"] = rate }
  @objc(setVolume:) func setVolume(_ volume: Double) { player.volume = Float(volume); lastState["volume"] = volume }
  @objc(setMuted:) func setMuted(_ muted: Bool) { player.isMuted = muted; lastState["muted"] = muted }
  @objc(setRepeatMode:) func setRepeatMode(_ mode: String) { repeatMode = mode }
  @objc(setShuffle:) func setShuffle(_ s: Bool) { shuffle = s; rebuildOrder() }

  @objc func setRemoteControls(_ optionsJson: String) {
    let opts = (try? JSONDecoder().decode(RemoteControlOptions.self, from: Data(optionsJson.utf8))) ?? RemoteControlOptions()
    remote.setEnabled(true, options: opts)
  }
  @objc func setBackgroundEnabled(_ enabled: Bool) {
    if enabled {
      configureSession()
      remote.setEnabled(true)
    } else {
      remote.unregister()
      try? AVAudioSession.sharedInstance().setActive(false)
    }
  }

  @objc func getCurrentState(_ resolve: @escaping RCTPromiseResolveBlock,
                             reject: @escaping RCTPromiseRejectBlock) {
    resolve(stateJson())
  }
  @objc func getCurrentQueue(_ resolve: @escaping RCTPromiseResolveBlock,
                             reject: @escaping RCTPromiseRejectBlock) {
    resolve(queueJson())
  }

  // MARK: - Internals

  private func configureSession() {
    let s = AVAudioSession.sharedInstance()
    try? s.setCategory(.playback, mode: .default, options: [.allowAirPlay, .allowBluetooth])
    try? s.setActive(true)
  }

  private func handleRemoteCommand(_ cmd: String) {
    switch cmd {
    case "play": player.play()
    case "pause": player.pause()
    case "next": next()
    case "previous": previous()
    default: break
    }
    emit("onRemoteCommand", ["command": cmd])
  }

  private func updateNowPlaying() {
    guard !order.isEmpty else { return }
    let t = tracks[order[cursor]]
    var info: [String: Any] = [:]
    if let title = t.title { info[MPMediaItemPropertyTitle] = title }
    if let artist = t.artist { info[MPMediaItemPropertyArtist] = artist }
    if let album = t.album { info[MPMediaItemPropertyAlbumTitle] = album }
    info[MPMediaItemPropertyPlaybackDuration] = player.currentItem?.duration.seconds ?? 0
    info[MPNowPlayingInfoPropertyElapsedPlaybackTime] = player.currentTime().seconds
    info[MPNowPlayingInfoPropertyPlaybackRate] = player.rate
    remote.updateNowPlaying(info)
  }

  private func decodeTracks(_ json: String) -> [MusicTrack] {
    (try? JSONDecoder().decode([MusicTrack].self, from: Data(json.utf8))) ?? []
  }
  private func trackJson(at index: Int) -> String {
    (try? String(data: JSONEncoder().encode(tracks[index]), encoding: .utf8)) ?? "{}"
  }
  private func queueJson() -> String {
    let simple: [String: Any] = ["index": currentIndex(), "ids": tracks.map { $0.id }]
    return (try? JSONSerialization.data(withJSONObject: simple)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
  }
  private func stateJson() -> String {
    (try? JSONSerialization.data(withJSONObject: lastState)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
  }
  private func emitQueue() {
    emit("onQueueChange", ["tracks": tracks.map { $0.id }, "index": currentIndex()])
  }
  private func delegateProgress() {
    emit("onProgress", ["position": lastState["position"] as? Double ?? 0,
                        "duration": lastState["duration"] as? Double ?? 0])
    emit("onState", ["stateJson": stateJson()])
  }

  override func observeValue(
    forKeyPath keyPath: String?, of object: Any?,
    change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?
  ) {
    guard let item = object as? AVPlayerItem else { return }
    if keyPath == #keyPath(AVPlayerItem.status) {
      switch item.status {
      case .readyToPlay:
        lastState["duration"] = item.duration.seconds
        lastState["status"] = player.timeControlStatus == .playing ? "playing" : "ready"
        updateNowPlaying()
        emit("onState", ["stateJson": stateJson()])
      case .failed:
        if let err = item.error { emit("onError", ["message": err.localizedDescription]) }
      default: break
      }
    } else if keyPath == #keyPath(AVPlayerItem.loadedTimeRanges) {
      if let range = item.loadedTimeRanges.last?.timeRangeValue { lastState["buffered"] = range.end.seconds }
    }
  }

  @objc private func itemEnded() {
    if repeatMode == "track" {
      player.seek(to: .zero); player.play()
    } else {
      next()
    }
  }
}

#if RCT_NEW_ARCH_ENABLED
import ObsidianMediaPlayerSpec
extension ObsidianMusicPlayer: ObsidianMusicPlayerSpec {}
#endif

File: android/src/main/java/com/obsidianmediaplayer/music/ObsidianMusicPlayerModule.kt
package com.obsidianmediaplayer.music

import android.content.Intent
import android.os.Build
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.session.MediaSession
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.obsidianmediaplayer.core.ExoPlayerProvider
import org.json.JSONArray
import org.json.JSONObject

/**
 * Full-featured music player with queue, shuffle/repeat, background audio
 * (foreground Service + MediaSession) and lock-screen remote controls.
 * Registered as "ObsidianMusicPlayer".
 */
class ObsidianMusicPlayerModule(private val ctx: ReactApplicationContext) : ReactContextBaseJavaModule(ctx) {
  override fun getName() = NAME

  companion object {
    const val NAME = "ObsidianMusicPlayer"
  }

  private val player: ExoPlayer by lazy { ExoPlayerProvider.buildPlayer(ctx) }
  private var mediaSession: MediaSession? = null

  private data class Track(
    val id: String,
    val uri: String,
    val headers: Map<String, String>,
    val title: String?,
    val artist: String?,
    val type: String?,
    val drmLicenseUri: String?,
    val cacheable: Boolean
  )

  private var tracks: MutableList<Track> = mutableListOf()
  private var order: MutableList<Int> = mutableListOf()
  private var cursor = 0
  private var repeatMode = "off"
  private var shuffle = false
  private val lastState = mutableMapOf<String, Any>(
    "status" to "idle", "position" to 0.0, "duration" to 0.0,
    "rate" to 1.0, "muted" to false, "volume" to 1.0,
    "buffered" to 0.0, "inBackground" to false
  )

  private val listener = object : Player.Listener {
    override fun onPlaybackStateChanged(state: Int) {
      lastState["status"] = when (state) {
        Player.STATE_READY -> if (player.playWhenReady) "playing" else "ready"
        Player.STATE_BUFFERING -> "buffering"
        Player.STATE_ENDED -> "ended"
        else -> "loading"
      }
      if (state == Player.STATE_ENDED) handleEnded()
      emitState()
    }
    override fun onIsPlayingChanged(isPlaying: Boolean) {
      lastState["status"] = if (isPlaying) "playing" else "paused"
      emitState()
    }
    override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
      send("onError", Arguments.createMap().apply { putString("message", error.message) })
    }
  }

  init { player.addListener(listener) }

  private fun send(name: String, params: WritableMap?) {
    ctx.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit(name, params)
  }
  private fun emitState() {
    send("onState", Arguments.createMap().apply { putString("stateJson", JSONObject(lastState as Map<*,*>).toString()) })
  }
  private fun emitQueue() {
    send("onQueueChange", Arguments.createMap().apply {
      putArray("tracks", Arguments.createArray().apply { tracks.forEach { pushString(it.id) } })
      putInt("index", currentIndex())
    })
  }

  @ReactMethod fun addListener(eventName: String) {}
  @ReactMethod fun removeListeners(count: Int) {}

  private fun currentIndex(): Int = if (order.isEmpty()) -1 else order[cursor]

  private fun rebuildOrder(preserve: Boolean) {
    val prevActive = if (preserve && order.isNotEmpty() && cursor in order.indices) order[cursor] else -1
    val n = tracks.size
    order = (0 until n).toMutableList()
    if (shuffle && n > 1) {
      order.shuffle()
      if (prevActive >= 0) {
        val pos = order.indexOf(prevActive)
        if (pos > 0) { val tmp = order[0]; order[0] = order[pos]; order[pos] = tmp }
        cursor = 0
      }
    }
    if (cursor >= order.size) cursor = 0
  }

  private fun decodeTracks(json: String): List<Track> {
    val arr = JSONArray(json); val out = mutableListOf<Track>()
    for (i in 0 until arr.length()) {
      val o = arr.getJSONObject(i)
      val src = o.getJSONObject("source")
      val headers = src.optJSONObject("headers")?.let { jo ->
        buildMap { jo.keys().forEach { k -> put(k, jo.getString(k)) } }
      } ?: emptyMap()
      val type = src.optString("type").takeIf { it.isNotEmpty() }
      val drm = src.optString("drmLicenseUri").takeIf { it.isNotEmpty() }
      val cacheable = if (src.has("cacheable")) src.optBoolean("cacheable", true) else true
      out += Track(o.getString("id"), src.getString("uri"), headers, o.optString("title").takeIf { it.isNotEmpty() }, o.optString("artist").takeIf { it.isNotEmpty() }, type, drm, cacheable)
    }
    return out
  }

  private fun loadCurrent(autoPlay: Boolean = false) {
    if (order.isEmpty()) return
    val idx = order[cursor]
    val t = tracks[idx]
    val item = MediaItem.Builder().setUri(t.uri).setMediaId(t.id).build()
    val source = ExoPlayerProvider.buildMediaSource(ctx, item, t.headers, t.type, t.cacheable, t.drmLicenseUri)
    player.setMediaSource(source); player.prepare()
    lastState["status"] = "loading"; emitState()
    send("onTrackChange", Arguments.createMap().apply { putInt("index", idx); putString("id", t.id) })
    if (autoPlay) player.play()
    ensureMediaSession()
  }

  private fun ensureMediaSession() {
    if (mediaSession == null) mediaSession = MediaSession.Builder(ctx, player).build()
  }

  @ReactMethod fun setQueue(tracksJson: String) {
    tracks = decodeTracks(tracksJson).toMutableList()
    rebuildOrder(false); cursor = 0; loadCurrent(); emitQueue()
  }
  @ReactMethod fun addTracks(tracksJson: String) { tracks.addAll(decodeTracks(tracksJson)); rebuildOrder(true); emitQueue() }
  @ReactMethod fun removeTrack(id: String) { tracks.removeAll { it.id == id }; rebuildOrder(true); emitQueue(); loadCurrent() }
  @ReactMethod fun skipTo(index: Int) {
    val pos = order.indexOf(index)
    cursor = if (pos >= 0) pos else index.coerceIn(0, maxOf(0, order.size - 1))
    loadCurrent()
  }
  @ReactMethod fun next() {
    if (order.isEmpty()) return
    if (cursor < order.size - 1) cursor++ else if (repeatMode == "queue") cursor = 0 else { send("onEnded", null); return }
    loadCurrent(true)
  }
  @ReactMethod fun previous() {
    if (order.isEmpty()) return
    cursor = if (cursor > 0) cursor - 1 else if (repeatMode == "queue") order.size - 1 else 0
    loadCurrent(true)
  }
  @ReactMethod fun play() { player.play() }
  @ReactMethod fun pause() { player.pause() }
  @ReactMethod fun stop() { player.pause(); player.seekTo(0) }
  @ReactMethod fun seek(seconds: Double) { player.seekTo((seconds * 1000).toLong()) }
  @ReactMethod fun setRate(rate: Double) {
    lastState["rate"] = rate
    player.setPlaybackParameters(androidx.media3.common.PlaybackParameters(rate.toFloat()))
  }
  @ReactMethod fun setVolume(volume: Double) { lastState["volume"] = volume; player.volume = volume.toFloat() }
  @ReactMethod fun setMuted(muted: Boolean) { lastState["muted"] = muted; player.volume = if (muted) 0f else (lastState["volume"] as Double).toFloat() }
  @ReactMethod fun setRepeatMode(mode: String) { repeatMode = mode }
  @ReactMethod fun setShuffle(s: Boolean) { shuffle = s; rebuildOrder(true) }
  @ReactMethod fun setRemoteControls(optionsJson: String) { ensureMediaSession() }
  @ReactMethod fun setBackgroundEnabled(enabled: Boolean) {
    if (enabled) {
      val intent = Intent(ctx, ObsidianPlaybackService::class.java)
      if (tracks.isNotEmpty()) {
        val t = tracks[currentIndex().coerceAtLeast(0).coerceAtMost(maxOf(0, tracks.size - 1))]
        intent.putExtra("title", t.title ?: t.id); intent.putExtra("artist", t.artist ?: "")
      }
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) ctx.startForegroundService(intent) else ctx.startService(intent)
      ensureMediaSession()
    } else {
      ctx.stopService(Intent(ctx, ObsidianPlaybackService::class.java))
    }
  }
  @ReactMethod fun getCurrentState(p: Promise) { p.resolve(JSONObject(lastState as Map<*,*>).toString()) }
  @ReactMethod fun getCurrentQueue(p: Promise) {
    val o = JSONObject().apply { put("index", currentIndex()); put("ids", JSONArray(tracks.map { it.id })) }
    p.resolve(o.toString())
  }

  private fun handleEnded() {
    if (repeatMode == "track") { player.seekTo(0); player.play() } else next()
  }

  override fun invalidate() { mediaSession?.release(); player.release(); super.invalidate() }
}
